### DISCUSSION TITLE
PSA: Improved Java Interoperability

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#8210] [Announcements]
We are pleased to announce that the next version of Flix will feature much
improved Java support. I will let the code speak for itself:

```scala
import java.lang.Math

def pythagoras(x: Float64, y: Float64): Float64 =
    unsafe Math.sqrt((Math.pow(x, 2.0) + Math.pow(y, 2.0)))

def main(): Unit \ IO = 
    println(pythagoras(3.0, 4.0))
```

```scala
import java.io.File

def main(): Unit \ IO = 
    let f = new File("foo.txt");
    if (f.exists())
        println("The file ${f.getName()} exists!")
    else
        println("The file ${f.getName()} does not exist!")
```

```scala
import java.io.File
import java.io.FileWriter
import java.io.IOException

def main(): Unit \ IO = 
    let f = new File("foo.txt");
    try {
        let w = new FileWriter(f);
        w.append("Hello World\n");
        w.close()
    } catch {
        case ex: IOException => 
            println("Unable to write to file: ${f.getName()}");
            println("The error message was: ${ex.getMessage()}")
    }
```

The documentation at https://doc.flix.dev/interoperability.html has been fully updated.

The Flix Standard Library has been refactored to use the new style.

To reduce breakage, we will continue to support the old style imports for
several months (or more). The `--Xno-deprecated` flag can be used to point out
usage of deprecated features. We suggest to migrate code within 1/2 year.

What's expected to work:

- Creating objects and calling object methods.
- Accessing static fields and calling static methods.
- Catching exceptions.

This includes support for overloaded constructors and methods.

What's coming shortly (next week):

- Seamless support for varargs.
- Support for throwing exceptions.
- Accessing object fields.

What's coming on a longer timeline (months):

- Writing object and static fields  (old syntax will remain in place).
- Seamless support for Java generics.
- Better error messages.
- Better auto-completion.

What's not supported:

- No autoboxing. Use Box.box and Box.unbox.
- Accessing inner non-static classes.

Most of the interop code is new, so we are looking for help to shake out bugs.

Thanks to the heroic efforts of @chanattan and the whole Flix team for making
this happen!

**Comment by magnus-madsen** [comment]
@JonathanStarup @mlutze @AStenbaek @chanattan @dghosef @jaschdoc @sockmaster27 @stephentetley CC

**Comment by magnus-madsen** [comment]
@stephentetley We would be pleased to know your experience with the new system (when you have time of course.)

**Comment by stephentetley** [comment]
Thanks Magnus, I've only done a simple interop project so far `charset-locale` but it was very easy to use and it cuts down the boilerplate nicely. I'm back to using Flix again (I had to do quite a bit of work related coding in Python) so I should be reactivating my broken libraries. I should be on with `flix-time` in a week or two, that's fairly simple as far as interop with Java goes but there are a lot more modules than `charset-locale`. 

The other libraries like `flix-basicdb` I'll be looking to use effects (vis "What if Java names were effects?" #8043) so they will be a more involved. I hope to be working on them by next month.
